src/storage/mvcc/reader/forward_scanner.rs

```diff
@@ -0,0 +1,459 @@
+                        // We need to further step write cursor to our desired version
+                        (Key::truncate_ts_for(k)?, true, false)
+                    }
+                    (Some(wk), Some(lk)) => {
+                        let write_user_key = Key::truncate_ts_for(wk)?;
+                        match write_user_key.cmp(lk) {
+                            Ordering::Less => {
+                                // Write cursor user key < lock cursor, it means the lock of the
+                                // current key that write cursor is pointing to does not exist.
+                                (write_user_key, true, false)
+                            }
+                            Ordering::Greater => {
+                                // Write cursor user key > lock cursor, it means we got a lock of a
+                                // key that does not have a write. In SI, we need to check if the
+                                // lock will cause conflict.
+                                (lk, false, true)
+                            }
+                            Ordering::Equal => {
+                                // Write cursor user key == lock cursor, it means the lock of the
+                                // current key that write cursor is pointing to *exists*.
+                                (lk, true, true)
+                            }
+                        }
+                    }
+                };
+
+                // Use `from_encoded_slice` to reserve space for ts, so later we can append ts to
+                // the key or its clones without reallocation.
+                (Key::from_encoded_slice(res.0), res.1, res.2)
+            };
+
+            // `result` stores intermediate values, including errors. When there is an error, for
+            // example, key locked error, we need to do the rest of the work like stepping cursors
+            // instead of stop there.
+            let mut result = Ok(None);
+
+            // `get_ts` is the real used timestamp. If user specifies `MaxInt64` as the timestamp,
+            // we need to change it to a most recently available one.
+            let mut get_ts = self.ts;
+
+            // `met_next_user_key` stores whether the write cursor has been already pointing to
+            // the next user key. If so, we don't need to compare it again when trying to step
+            // to the next user key later.
+            let mut met_next_user_key = false;
+
+            if has_lock {
+                match self.isolation_level {
+                    IsolationLevel::SI => {
+                        assert!(self.lock_cursor.valid());
+                        let lock = {
+                            let lock_value = self.lock_cursor.value(&mut self.statistics.lock);
+                            Lock::parse(lock_value)?
+                        };
+                        match super::util::check_lock(&current_user_key, self.ts, &lock) {
```

Comment:
You can return error directly when check lock failed.

Reply:
We cannot do this for now, because we need to continue the remaining steps (move cursor to next / prev user key) when meet key lock errors in kv scan interface.

Reply:
Got it.